File: closure/src/codekata/chop/chop.clj
(ns codekata.chop.chop)


(defn split-in-middle [xs]
  (split-at (int (/ (count xs) 2)) xs)
  )

;; Recursive binary search
(defn chop
  ;; Mehhod is called first time
  ([x xs]
    ;; 0 is initial offset
    (chop x xs 0))
  ([x xs offset]
    (do
      (prn x xs offset)
      (cond
        ;; If list is empty return -1
        (empty? xs) -1
        ;; If element x is the last element in the list calculate and return index
        (== x (last xs)) (+ offset (- (count xs) 1))
        ;; If list contains only one element return -1
        (== 1 (count xs)) -1
        :else ;; Split list in the middle
        (let [parts (split-in-middle xs) parts_head (first parts) parts_tail (last parts)]
          ;; If the last element of first list is less than or equal x then search in the first list otherwise in the second
          (if (<= x (last parts_head))
            (recur x parts_head offset)
            (recur x parts_tail (+ offset (count parts_head)))
            )
          )
        )
      )
    )
  )
;;(prn "Result" (chop 2 '(1 2 3)))

File: python/codekata/chop.py
__author__ = 'mehmet'

class BinarySearch:
    @staticmethod
    def split_list(a_list):
        half = int(len(a_list)/2)
        print(a_list, half)
        return a_list[:half], a_list[half:]

    @staticmethod
    def chop(element, data, offset=0):
        # if list is empty return -1
        print(element, data, offset)
        if(not data):
          return -1
        # If element is same as the last element in list then calculate and return index
        if element == data[-1]:
            return offset + len(data) -1
        # If size of the data is one then return -1
        if len(data) == 1:
            return -1
        head, tail=BinarySearch.split_list(data)
        if element<= head[-1]:
            return BinarySearch.chop(element=element, data=head, offset=offset)
        else:
            return BinarySearch.chop(element=element, data=tail, offset=offset + len(head))